// @tbranyen/jsdom/lib/jsdom/living/generated/CustomElementRegistry.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");

const convertElementDefinitionOptions = require("./ElementDefinitionOptions.js").convert;
const convertNode = require("./Node.js").convert;
const impl = utils.implSymbol;

class CustomElementRegistry {
  constructor() {
    throw new TypeError("Illegal constructor");
  }

  define(name, constructor) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    }

    if (arguments.length < 2) {
      throw new TypeError(
        "Failed to execute 'define' on 'CustomElementRegistry': 2 arguments required, but only " +
          arguments.length +
          " present."
      );
    }
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["DOMString"](curArg, {
        context: "Failed to execute 'define' on 'CustomElementRegistry': parameter 1"
      });
      args.push(curArg);
    }
    {
      let curArg = arguments[1];
      curArg = utils.tryImplForWrapper(curArg);
      args.push(curArg);
    }
    {
      let curArg = arguments[2];
      curArg = convertElementDefinitionOptions(curArg, {
        context: "Failed to execute 'define' on 'CustomElementRegistry': parameter 3"
      });
      args.push(curArg);
    }
    return this[impl].define(...args);
  }

  get(name) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    }

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'get' on 'CustomElementRegistry': 1 argument required, but only " +
          arguments.length +
          " present."
      );
    }
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["DOMString"](curArg, {
        context: "Failed to execute 'get' on 'CustomElementRegistry': parameter 1"
      });
      args.push(curArg);
    }
    return this[impl].get(...args);
  }

  whenDefined(name) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    }

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'whenDefined' on 'CustomElementRegistry': 1 argument required, but only " +
          arguments.length +
          " present."
      );
    }
    const args = [];
    {
      let curArg = arguments[0];
      curArg = conversions["DOMString"](curArg, {
        context: "Failed to execute 'whenDefined' on 'CustomElementRegistry': parameter 1"
      });
      args.push(curArg);
    }
    return utils.tryWrapperForImpl(this[impl].whenDefined(...args));
  }

  upgrade(root) {
    if (!this || !module.exports.is(this)) {
      throw new TypeError("Illegal invocation");
    }

    if (arguments.length < 1) {
      throw new TypeError(
        "Failed to execute 'upgrade' on 'CustomElementRegistry': 1 argument required, but only " +
          arguments.length +
          " present."
      );
    }
    const args = [];
    {
      let curArg = arguments[0];
      curArg = convertNode(curArg, { context: "Failed to execute 'upgrade' on 'CustomElementRegistry': parameter 1" });
      args.push(curArg);
    }
    return this[impl].upgrade(...args);
  }
}
Object.defineProperties(CustomElementRegistry.prototype, {
  define: { enumerable: true },
  get: { enumerable: true },
  whenDefined: { enumerable: true },
  upgrade: { enumerable: true },
  [Symbol.toStringTag]: { value: "CustomElementRegistry", configurable: true }
});
const iface = {
  // When an interface-module that implements this interface as a mixin is loaded, it will append its own `.is()`
  // method into this array. It allows objects that directly implements *those* interfaces to be recognized as
  // implementing this mixin interface.
  _mixedIntoPredicates: [],
  is(obj) {
    if (obj) {
      if (utils.hasOwn(obj, impl) && obj[impl] instanceof Impl.implementation) {
        return true;
      }
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(obj)) {
          return true;
        }
      }
    }
    return false;
  },
  isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      }

      const wrapper = utils.wrapperForImpl(obj);
      for (const isMixedInto of module.exports._mixedIntoPredicates) {
        if (isMixedInto(wrapper)) {
          return true;
        }
      }
    }
    return false;
  },
  convert(obj, { context = "The provided value" } = {}) {
    if (module.exports.is(obj)) {
      return utils.implForWrapper(obj);
    }
    throw new TypeError(`${context} is not of type 'CustomElementRegistry'.`);
  },

  create(constructorArgs, privateData) {
    let obj = Object.create(CustomElementRegistry.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return obj;
  },
  createImpl(constructorArgs, privateData) {
    let obj = Object.create(CustomElementRegistry.prototype);
    obj = this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  },
  _internalSetup(obj) {},
  setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {};

    privateData.wrapper = obj;

    this._internalSetup(obj);
    Object.defineProperty(obj, impl, {
      value: new Impl.implementation(constructorArgs, privateData),
      configurable: true
    });

    obj[impl][utils.wrapperSymbol] = obj;
    if (Impl.init) {
      Impl.init(obj[impl], privateData);
    }
    return obj;
  },
  interface: CustomElementRegistry,
  expose: {
    Window: { CustomElementRegistry }
  }
}; // iface
module.exports = iface;

const Impl = require("../custom-elements/CustomElementRegistry-impl.js");
